Migration docs (#28)
signalfx/splunk-otel-ruby: MIGRATING.md

# Migrate from the SignalFx Tracing Library for Ruby

> The SignalFx Tracing Library for Ruby will soon be deprecated. Replace it with the

Reviewer: I think this is too vague. Do we have any dates for this? Can we say something like "will be deprecated by the end of 2022", for instance?

# Migrate from the SignalFx Tracing Library for Ruby

The SignalFx Tracing Library for Ruby is scheduled to be deprecated by the end of 2022.

> Replace it with the agent from the Splunk Distribution of OpenTelemetry Ruby.

Reviewer 2: I don't think we should use the term "agent" here and in other places in this document that aren't part of this PR. Just call it the Distribution or a library?

The Splunk Distribution of OpenTelemetry Ruby requires Ruby 2.6 and
higher.

## Known limitations compared to SignalFx Tracing for Ruby

- Different subset of supported Ruby versions, see previous section

> - No auto-instrumentation for
>   (more details in
>   [OTel instrumentation equivalents](#signalfx-instrumentations-equivalents)
>   ):
>     - `elasticsearch` (<https://github.com/elastic/elasticsearch-ruby>)
>     - `grape` (<https://github.com/ruby-grape/grape>)
>     - `sequel` (<https://github.com/jeremyevans/sequel>)
> 

Reviewer 2: ```suggestion
- Currently no auto-instrumentation for:
    - `elasticsearch` (<https://github.com/elastic/elasticsearch-ruby>)
    - `grape` (<https://github.com/ruby-grape/grape>)
    - `sequel` (<https://github.com/jeremyevans/sequel>)

More details in [OTel instrumentation equivalents](#signalfx-instrumentations-equivalents).
```

    gem uninstall signalfx
    ```

> 1.  Remove `signalfx` from your Gemfile.

Reviewer 2: ```suggestion
2.  Remove `signalfx` from your Gemfile.
```

### Deploy the Splunk Ruby agent

To install the Splunk Distribution of OpenTelemetry Ruby, see the [README.md](README.md).

## Migrate to the Splunk Distribution of OpenTelemetry Ruby

We're currently only testing installation using Bundler with RubyGems as a source.
[Contact us](mailto:ssg-observability-instrumentals-ruby@splunk.com) if you require a different way of installation.

If you installed
[Smart Agent](https://github.com/signalfx/signalfx-agent)
to serve as a gateway (traces/metrics data proxy)
migrate to
[OpenTelemetry Collector](https://docs.splunk.com/Observability/gdi/opentelemetry/resources.html)
first, as soon as (if) possible.

### Deploy the Splunk Ruby agent

To install the Splunk Distribution of OpenTelemetry Ruby, see the [README.md](README.md).

### Replace SignalFx / OpenTracing gems with OTel equivalents

1.  Make a list of all `signalfx-*` instrumentation gems from your Gemfile, see
    [this table](https://github.com/signalfx/signalfx-ruby-tracing#supported-libraries)
    for a complete list.

> 1.  Replace them with Open Telemetry, as per the table below.

Reviewer 2: ```suggestion
1.  Replace them with OpenTelemetry, as per the table below.
```

> 1.  Replace any other OpenTracing instrumentation packages you might have installed yourself.

Reviewer 2: ```suggestion
2.  Replace any other OpenTracing instrumentation packages you might have installed yourself.
```

1. `SIGNALFX_SPAN_TAGS` or `span_tags` configuration parameter
    - to set in environment:
        - first, you need to rewrite the existing value to the,
          which has a syntax of comma-separated `<key>=<value>` pairs (e.g. `key1=value1,key2=value2`)
        - set `OTEL_RESOURCE_ATTRIBUTES` to the new value
    - to set in code:
        - call `OpenTelemetry::SDK::Resources::Resource.create` with a `Hash`
          consisting of your existing key-value pairs
        - pass the newly created `Resource` to the `resource=` attribute in configuration
        - for example:
           ```ruby
           Splunk::Otel.configure(other_args) do |c|
             c.resource = OpenTelemetry::SDK::Resources::Resource.create(
               "key" => "value"
             )
             # ...
           end
           ```

1. `SIGNALFX_RECORDED_VALUE_MAX_LENGTH`
    - can be set in environment using `OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT`
    - see
      [the list of span limits settings](<https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#span-limits>)
      to learn more

> 1. `tracer` configuration parameter

Reviewer 2: ```suggestion
2. `tracer` configuration parameter
```

    gem uninstall signalfx
    ```

1.  Remove `signalfx` from your Gemfile.

### Deploy the Splunk Ruby agent

To install the Splunk Distribution of OpenTelemetry Ruby, see the [README.md](README.md).

## Migrate to the Splunk Distribution of OpenTelemetry Ruby

We're currently only testing installation using Bundler with RubyGems as a source.
[Contact us](mailto:ssg-observability-instrumentals-ruby@splunk.com) if you require a different way of installation.

If you installed
[Smart Agent](https://github.com/signalfx/signalfx-agent)
to serve as a gateway (traces/metrics data proxy) then,
if possible, migrate to
[OpenTelemetry Collector](https://docs.splunk.com/Observability/gdi/opentelemetry/resources.html)

> first.

Reviewer: Suggestion: If you installed [Smart Agent](https://github.com/signalfx/signalfx-agent)
to serve as a gateway (traces/metrics data proxy), migrate to
[OpenTelemetry Collector](https://docs.splunk.com/Observability/gdi/opentelemetry/resources.html)
first, as soon as (if) possible.

### Deploy the Splunk Ruby agent

To install the Splunk Distribution of OpenTelemetry Ruby, see the [README.md](README.md).

### Replace SignalFx / OpenTracing gems with OTel equivalents

1.  Make a list of all `signalfx-*` instrumentation gems from your Gemfile, see
    [this table](https://github.com/signalfx/signalfx-ruby-tracing#supported-libraries)
    for a complete list.

> 1.  Replace them with Open Telemetry equivalents per the table below.

Reviewer: Numbering.

Reviewer: Replace them with Open Telemetry equivalents, as per the table below.

| signalfx-sidekiq-opentracing           | opentelemetry-instrumentation-sidekiq | |
| signalfx-sinatra-instrumentation       | opentelemetry-instrumentation-sinatra | |

#### OTel equivalents of other Open Tracing instrumentation gems

See the list of all known OTel Ruby instrumentations: <https://opentelemetry.io/registry/?language=ruby&component=instrumentation>

### Migrate settings for the Splunk Ruby OTel agent

1. `SIGNALFX_ENDPOINT_URL` or `ingest_url` configuration parameter
    - if you installed Smart Agent and can migrate to OpenTelemetry Collector, do that first, then see the point below
    - if you have OpenTelemetry Collector available up as a sidecar (via `localhost`),
      and it accepts OTLP on default ports, then just remove this setting, we export OTLP to OTel Collector by default
    - if you need to keep using Smart Agent, then you need to set up a jaeger exporter yourself
    - if you export directly to our backend (without OTel Collector as a proxy),
      then just set `SPLUNK_REALM` to your realm

> 1. `SIGNALFX_SERVICE_NAME` or `service_name` configuration parameter

Reviewer: Numbering

>     - to set in environment use `OTEL_SERVICE_NAME`

Reviewer: to set in environment, use `OTEL_SERVICE_NAME`

    - to set in code use `Splunk::Otel.configure(service_name: 'your service name', ...`

> 

Reviewer: to set in code, use `Splunk::Otel.configure(service_name: 'your service name', ...`

1. `SIGNALFX_ACCESS_TOKEN` or `access_token` configuration parameter
    - can only be set in environment using `SPLUNK_ACCESS_TOKEN`

1. `SIGNALFX_SPAN_TAGS` or `span_tags`

>     - TODO: check if we provide a way set resource

Author: I'll be looking into this next, I suggest that we just keep this TODO here until I get it resolved

1. `SIGNALFX_RECORDED_VALUE_MAX_LENGTH`
    - can be set in environment using `OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT`

>     - see <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#span-limits>

Reviewer: Add a description for the url
